src/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+import json
+import logging
+import random
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Dict, List, overload
+
+from pytorch_ie.annotations import BinaryRelation, LabeledSpan, Span
+from pytorch_ie.core import AnnotationList, annotation_field
+from pytorch_ie.documents import TextDocument
+from pytorch_ie.utils.span import is_contained_in
+
+from pie_utils.span.slice import distance
+from pie_utils.statistics import WithStatistics
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class DocumentWithEntitiesRelationsAndPartition(TextDocument):
+    entities: AnnotationList[LabeledSpan] = annotation_field(target="text")
+    relations: AnnotationList[BinaryRelation] = annotation_field(target="entities")
+    partition: AnnotationList[Span] = annotation_field(target="text")
+
+
+class CandidateRelationAdder(WithStatistics):
+    def reset_statistics(self):
+        self._statistics: dict[str, Any] = {
+            "num_total_relation_candidates": 0,
+            "num_available_relations": 0,
+            "available_rels_within_allowed_distance": 0,
+            "num_added_relation_not_taken": 0,
+            "num_rels_within_allowed_distance": 0,
+            "num_candidates_not_taken": defaultdict(int),
+            "distances_taken": defaultdict(list),
+        }
+
+    def show_statistics(self, description: str | None = None):
+        description = description or "Statistics"
+        logger.info(f"{description}: \n{json.dumps(self._statistics, indent=2)}")
+
+    @overload
+    def update_statistics(self, key: str, value: int):
+        pass
```

Comment:
why is this empty? (also below)

Reply:
clarified in https://github.com/ArneBinder/pie-utils/pull/14#discussion_r938762791